### PR TITLE
Improve test coverage for public models

### PR DESCRIPTION
**Pull request by joeydong-stripe**
## Summary

Improves test coverage for our public models and improves some implementations of `STPFormEncodable`

## Motivation

Better test coverage should help us iterate with more confidence.

## Testing

Verified unit tests still pass.

r? @bdorfman-stripe 
cc @bg-stripe

### CODE REVIEW

File: Tests/Tests/STPBankAccountTest.m [hunk 1]
```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
...
+
+    XCTAssertEqualObjects(bankAccount.bankAccountId, @"ba_1AXvnKEOD54MuFwSotKc6xq0");
+    XCTAssertEqualObjects(bankAccount.accountHolderName, @"Jane Austen");
+    XCTAssertEqual(bankAccount.accountHolderType, STPBankAccountHolderTypeIndividual);
+    XCTAssertEqualObjects(bankAccount.bankName, @"STRIPE TEST BANK");
+    XCTAssertEqualObjects(bankAccount.country, @"US");
+    XCTAssertEqualObjects(bankAccount.currency, @"usd");
+    XCTAssertEqualObjects(bankAccount.fingerprint, @"C5fW7AwE3of8bHvV");
+    XCTAssertEqualObjects(bankAccount.last4, @"6789");
+    XCTAssertNil(bankAccount.routingNumber);
```
Comment (joeydong-stripe) [inline Tests/Tests/STPBankAccountTest.m]: This one's interesting. Should we map the routingNumber to the bankAccount?

**Comment by joeydong-stripe** [comment]
Sorry for the commit churn. Also, I can break this PR up if this is too much at once.

File: Stripe/STPBankAccountParams+Private.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ NS_ASSUME_NONNULL_BEGIN
 + (STPBankAccountHolderType)accountHolderTypeFromString:(NSString *)string;
 + (NSString *)stringFromAccountHolderType:(STPBankAccountHolderType)accountHolderType;
 
+- (instancetype)init;
```
Comment (bdorfman-stripe) [inline Stripe/STPBankAccountParams+Private.h]: Does this need to be here? Don't all NSObject's inherit init anyway?

File: Tests/Tests/STPBankAccountTest.m [hunk 3]
```diff
@@ -8,21 +8,20 @@
...
+@interface STPBankAccountTest : XCTestCase
 
 @end
 
 @implementation STPBankAccountTest
 
 - (void)setUp {
     [super setUp];
-    _bankAccount = [[STPBankAccount alloc] init];
+    // Put setup code here. This method is called before the invocation of each test method in the class.
```
Comment (bdorfman-stripe) [inline Tests/Tests/STPBankAccountTest.m]: If there is no setup needed, you should just delete the method.

File: Tests/Tests/STPBankAccountTest.m [hunk 4]
```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
...
-- (void)testDescriptionWorks {
+- (void)testDescription {
     STPBankAccount *bankAccount = [STPBankAccount decodedObjectFromAPIResponse:[self completeAttributeDictionary]];
-    bankAccount.routingNumber = @"123456789";
     XCTAssert(bankAccount.description);
 }
 
 #pragma mark - STPAPIResponseDecodable Tests
 
 - (NSDictionary *)completeAttributeDictionary {
```
Comment (bdorfman-stripe) [inline Tests/Tests/STPBankAccountTest.m]: Test json like this should probably be moved to STPFixtures? And/or an actual json file.

File: Tests/Tests/STPBankAccountTest.m (hunk 1 above)
Comment (bdorfman-stripe) [inline Tests/Tests/STPBankAccountTest.m]: Yeah it seems like an oversight that this isn't parsed currently? Not sure if there was some intentional reason we don't parse this (although I assume not because we have a property for it). Also `routingNumber` is declared as nonnull, which is just entirely wrong. Should change that annotation and/or parse it and add it to required keys I guess.

Somewhat relatedly: It's suuuuper weird to me that STPBankAccount and STPCard are subclasses of their own Params objects. Glad we didn't do that for Sources.

What do you think @bg-stripe?

File: Tests/Tests/STPCardTest.m [hunk 5]
```diff
@@ -13,15 +13,13 @@
...
-@property (nonatomic) STPCard *card;
-
 @end
 
 @implementation STPCardTest
 
 - (void)setUp {
     [super setUp];
-    _card = [[STPCard alloc] init];
+    // Put setup code here. This method is called before the invocation of each test method in the class.
```
Comment (bdorfman-stripe) [inline Tests/Tests/STPCardTest.m]: Same as other setup comment.

File: Tests/Tests/STPCardTest.m [hunk 6]
```diff
@@ -151,38 +149,86 @@ - (void)testStringFromFunding {
...
+- (void)testInit {
+    STPCard *card = [[STPCard alloc] init];
+    XCTAssertEqual(card.brand, STPCardBrandUnknown);
+    XCTAssertEqual(card.funding, STPCardFundingTypeOther);
+}
+
 - (void)testLast4ReturnsCardNumberLast4WhenNotSet {
-    self.card.number = @"4242424242424242";
-    XCTAssertEqualObjects(self.card.last4, @"4242", @"last4 correctly returns the last 4 digits of the card number");
+    STPCard *card = [[STPCard alloc] init];
```
Comment (bdorfman-stripe) [inline Tests/Tests/STPCardTest.m]: can just do `[STPCard new]` instead of alloc/init? Or is that a style guide holy war? :)

File: Tests/Tests/STPCardTest.m [hunk 7]
```diff
@@ -191,66 +237,118 @@ - (void)testCardEquals {
...
-             @"address_line1": @"221A Baker Street",
-             @"address_city": @"New York",
-             @"address_state": @"NY",
-             @"address_zip": @"12345",
-             @"address_country": @"USA",
-             @"last4": @"1234",
-             @"dynamic_last4": @"5678",
-             @"brand": @"MasterCard",
-             @"country": @"Japan",
+             @"id": @"card_1AVRojEOD54MuFwSxr93QJSx",
```
Comment (bdorfman-stripe) [inline Tests/Tests/STPCardTest.m]: Where do these ids actually come from? I have different ones on my screen. Are we sure these aren't "real" tokens that are populated from the logged in Stripe account or something?

Also again I think all fake json stuff should be through STPFixtures now.

File: Tests/Tests/STPBankAccountTest.m (hunk 1 above)
Comment (bg-stripe) [inline Tests/Tests/STPBankAccountTest.m]: Yeah it is super weird – I _think_ the history there is that previously `STPCard` was used as both the params object and the returned object, and we wanted to introduce `STPCardParams` with a deprecation, rather than a breaking change.

I think the best fix here would be to make `STPBankAccount` not a subclass of `STPBankAccountParams` and do the same with `CardParams` – maybe we should save that for a separate PR though.

File: Tests/Tests/STPCardTest.m (hunk 6 above)
Comment (bg-stripe) [inline Tests/Tests/STPCardTest.m]: :P +1 for `new` in the style guide

File: Tests/Tests/STPCardTest.m (hunk 7 above)
Comment (bg-stripe) [inline Tests/Tests/STPCardTest.m]: (I usually use dummy IDs like `card_123` in fixtures)

File: Tests/Tests/STPBankAccountTest.m (hunk 1 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPBankAccountTest.m]: Cool, I'll at least map the `routingNumber`. I think overall `STPBankAccount` has a few nullability bugs that we can fix (cause public api changes) but I'm guessing the object isn't used very often so we don't hear much about it.

Are we ever using the fact that `STPBankAccount` is a subclass of `STPBankAccountParams` to allow people to retrieve/modify/resubmit a bank account easily? Maybe that's too magical.

File: Stripe/STPBankAccountParams+Private.h (hunk 2 above)
Comment (joeydong-stripe) [inline Stripe/STPBankAccountParams+Private.h]: I think I was getting at making sure we have an explicit designated initializer for everything and seeing if that provides any compiler safety but didn't finish it. I'll remove this for now and restart the investigation later.

File: Tests/Tests/STPCardTest.m (hunk 6 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPCardTest.m]: I added this as one of our style guide discussion points 😉

File: Tests/Tests/STPCardTest.m (hunk 7 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPCardTest.m]: haha yeah looks like it's one of my dummy test cards. i'll figure something out.

File: Tests/Tests/STPCardTest.m [hunk 8]
```diff
@@ -151,106 +142,176 @@ - (void)testStringFromFunding {
...
+    XCTAssertEqualObjects(card.currency, @"usd");
+    XCTAssertEqualObjects(card.dynamicLast4, @"5678");
+    XCTAssertEqual(card.expMonth, (NSUInteger)5);
+    XCTAssertEqual(card.expYear, (NSUInteger)2017);
+    XCTAssertEqual(card.funding, STPCardFundingTypeCredit);
+    XCTAssertEqualObjects(card.last4, @"4242");
+    XCTAssertEqualObjects(card.name, @"Jane Austen");
+
+    XCTAssertNotEqual(card.allResponseFields, response);
+    XCTAssertEqualObjects(card.allResponseFields, [response stp_dictionaryByRemovingNullsValidatingRequiredFields:@[]]);
```
Comment (joeydong-stripe) [inline Tests/Tests/STPCardTest.m]: I'll come back to fixing this `stp_dictionaryByRemovingNullsValidatingRequiredFields` usage in another PR. The category is missing test coverage and we should probably deeply remove `[NSNull null]`s.

**Comment by joeydong-stripe** [comment]
wdyt @bdorfman-stripe ?

File: Tests/Tests/STPBankAccountTest.m (hunk 3 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPBankAccountTest.m]: 👍

File: Tests/Tests/STPBankAccountTest.m (hunk 4 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPBankAccountTest.m]: 👍

File: Tests/Tests/STPCardTest.m (hunk 5 above)
Comment (joeydong-stripe) [inline Tests/Tests/STPCardTest.m]: 👍

**Comment by joeydong-stripe** [comment]
Just moved all of the private header definitions I added that were specific to tests in 0a89c95 because it looks like that's where our style is headed.

cc @bg-stripe
